### ISSUE TITLE
Document "fullTopologyInLastLevel" better and/or emit warning messages

### ISSUE THREAD

**Issue by kjym3** [PixarAnimationStudios/OpenSubdiv#764]
The GetEdgeVertices method of a Far::TopologyLevel instance at the maximum level of a uniform refinement returns an invalid Far::ConstIndexArray.  Specifically, access to the elements of the returned array leads to a crash due to out-of-range vector subscription.

A test case for reproducing the documented issue is found at https://gist.github.com/kjym3/5d7c48be7a7f91e54b03.  The code was tested using osdCPU.lib of the OpenSubdiv 3.0.3 release, compiled with VS Express 2013 for Windows Desktop.

**Comment by barfowl** [comment]
I looked at the test case, and while I have not verified it yet, I believe the issue here is the options that are set for the uniform refinement via the UniformOptions struct (at line 57).

There is a "fullTopologyInLastLevel" option of the UniformOptions struct that is set to false by default.  This option minimizes how much topological information is computed for the last level of refinement, as in many cases you only need the face list.  If you want to traverse the topology of the last level with the same flexibility as all others, you must explicitly set this option to true.  The memory savings of suppressing this information is considerable with larger meshes and higher levels of subdivision (given the geometric growth with uniform subdivision), which is why it was set to false by default.

So in your test case, there are no edge-vertices created at the maximum level for you to inspect.  There is some information for edges (e.g. the refined sharpness values), which is why there is a non-zero number of edges, but all other information relating to them is suppressed.

Declaring an instance of UniformOptions and setting "fullTopologyInLastLevel" to true should resolve this.

**Comment by jcowles** [comment]
Thanks Barry! We've hit this misuse a few times now, I wonder if we could make the issue more explicit/apparent to the user, perhaps with an error message from the appropriate functions -- what do you think? Is there a clean way to do something like this?

--
Jeremy

**Comment by barfowl** [comment]
Agreed.  Other than better documentation and maybe more highlighting of it in the tutorials, I don't have any compelling ideas.

We can't generate error messages for these two choices because both have legitimate use cases:  you want to use "false" for fast and lean uniform refinement when you only care about a face list and refined vertex values, but you want it "true" if you want to do more with the refined results.  The current single option was simplified down from originally wanting to be able to specify exactly what was to be refined (e.g. just the face-vertices and edge-vertices) which was really something for expert users.

We could put some tests and error messages in debug mode within the topology queries, e.g. GetEdgeVertices() et al.  That would help for those that actually make use of debug mode but not optimized (and I wouldn't want to slow down the lightweight topology queries with such runtime error checking in general).

**Comment by kjym3** [comment]
Many thanks for the suggested change to the use of UniformOptions.  It is confirmed that the test code now works as expected by changing the option value as follows:

```C++
Far::TopologyRefiner::UniformOptions uniformOptions(maxlevel);
uniformOptions.fullTopologyInLastLevel = true;
refiner->RefineUniform(uniformOptions);
```

**Comment by barfowl** [comment]
Great, thanks for confirming that all works as expected once the option is set.  Jeremy, assuming we can close this issue, do you think its worth opening a separate one to try and better communicate this in general?

**Comment by jcowles** [comment]
How about we just retitle this issue, since it already has all the state?

**Comment by salvager2014** [comment]
all right that be ok. you can do this. 发自网易邮箱大师 On 2015-12-14 15:42 , Jeremy Cowles Wrote: How about we just retitle this issue, since it already has all the state? — Reply to this email directly or view it on GitHub.

**Comment by jtran56** [comment]
Filed as internal issue #151661.

**Comment by barfowl** [comment]
This was addressed in version 3.1 with the commits noted above.
